### PR TITLE
Optimizer selector

### PR DESCRIPTION
**Pull request by lm-konda**
Implement optimizer selector.

## Motivation and Context
This PR add the functionality to select optimizer from Adam and Momentum. See issue #138 .

## How has this been tested?

I did experiments with `AdamOtimizer` and `MomentumOptimizer` with classification using CIFAR10 dataset. Here is the result.

https://github.com/blue-oil/blueoil/pull/28#issuecomment-457097443

I also added test case to `blueoil_test.sh`. We use Adam for Classification and Momentum for Object Detection.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by ruimashita** [comment]
Ok to test

**Comment by ruimashita** [comment]
Thanks, so good contribution 😄😄😄
Please add description about this feature in the `docs/usage/init.md` page.

File: blueoil/templates/lmnet/classification.tpl.py [hunk 1]
```diff
@@ -80,7 +80,13 @@
...
+
+if '{{optimizer}}' == 'GradientDescentOptimizer':
+    NETWORK.OPTIMIZER_CLASS = tf.train.GradientDescentOptimizer
+elif '{{optimizer}}' == 'MomentumOptimizer':
+    NETWORK.OPTIMIZER_CLASS = tf.train.MomentumOptimizer
+elif '{{optimizer}}' == 'AdamOptimizer':
+    NETWORK.OPTIMIZER_CLASS = tf.train.AdamOptimizer
 
 if '{{learning_rate_setting}}' != 'fixed':
     NETWORK.OPTIMIZER_KWARGS = {"momentum": 0.9}
```
Comment (ruimashita) [inline blueoil/templates/lmnet/classification.tpl.py]: Maybe `AdamOptimizer` and `GradientDescentOptimizer` don't accept `{"momentum": 0.9}`, but I'm not sure.

https://www.tensorflow.org/api_docs/python/tf/train/GradientDescentOptimizer
https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer

I will verify it's work well when optimizer is `GradientDescentOptimizer` or `AdamOptimizer`

**Comment by lm-konda** [comment]
Sorry, this is my mistake when I update branch.
I will fix it.

**Comment by lm-konda** [comment]
I fixed it and tested.

Training with SGD, Momentum and Adam optimizer seems to be working well.

**Comment by lm-konda** [comment]
Yes, of course I tested. But there may be missing.

**Comment by lm-konda** [comment]
Sorry, some commits were missing.

**Comment by lm-konda** [comment]
`All tests are passed!`

**Comment by ruimashita** [comment]
@lm-konda Please solve conflict ☺️

**Comment by lm-konda** [comment]
@ruimashita I solved conflicts.

File: tests/config/make_yml_config.py [hunk 2]
```diff
@@ -243,9 +251,9 @@ def learning_settings_to_yaml(index):
     # trainer epochs
     fp.write(str(trainer_epochs[index]))
     # trainer optimizer comment
-    #fp.write(str(trainer_optimizer_comment))
+    fp.write(str(trainer_optimizer_comment))
```
[the remark this answers is not in the document]
Comment (lm-konda) [reply tests/config/make_yml_config.py]: I think OK. But I will confirm it. please wait for a while.

[the remark this answers is not in the document]
Comment (lm-konda) [reply tests/config/make_yml_config.py]: I confirmed it.
And theoretically, blueoil_test.sh compares
the config generated from the template with the test config.
The template contains comments, so it is OK.

**Comment by iizukak** [comment]
## Cifar10 Training Test Result

I tested each optimizers wih Cifar10 Classification. And here is the result.
 I dont's know it reasonable result. But they are working, anyway.

### MomentumOptimizer

cifar10_momentum

### AdamOptimizer

cifar10_adam

### GradientDescentOptimizer

cifar10_sgd

**Comment by iizukak** [comment]
@ruimashita 
I added loss, metrics and final detect boxes screenshot images of Tensorboard.
And updated `blueoil.sh init` parameter because of master brach updated.

Please check
https://deploy-preview-33--blueoil.netlify.com/tutorial/image_det.html

**Comment by iizukak** [comment]
@ruimashita I added test case and documentation. Please review 🙇

File: docs/usage/init.md [hunk 3]
```diff
@@ -51,3 +51,21 @@ common:
 ```
 
 Please see <a href="../reference/data_augmentor.html">the data augmentor reference page</a>, when you want to know about augmentation methods. the all of augmentation methods and parameter are explained, methods name in generated yaml correspond to class name under `lmnet.data_augmentor` in the reference.
+
+
+#### optimizer
+
+You can choose optimizer from Adam or Momentum. Each optimizer uses TensorFlow implementation. Please see TensorFlow documentation, [AdamOptimizer](https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer) and [MomentumOptimizer](https://www.tensorflow.org/api_docs/python/tf/train/MomentumOptimizer).
```
Comment (ruimashita) [inline docs/usage/init.md]: Which optimizer can user use in config yaml, `Adam` or `AdamOptimizer` ? The document looks to use `Adam`, but actually `AdamOptimizer`.
Please change document more clearly what optimizer user can use 👍 

I prefer `Adam` because it is obvious `Adam` is optimizer. But It takes cost to change, no need to change I think 😭

Comment (iizukak) [reply docs/usage/init.md]: `Adam` is 👌

**Comment by iizukak** [comment]
@ruimashita Thank you comment! I fixed.

documentation
https://deploy-preview-28--blueoil.netlify.com/usage/init.html

**Comment by iizukak** [comment]
@ruimashita Please review this PR agein.

File: blueoil/generate_lmnet_config.py [hunk 4]
```diff
@@ -154,10 +162,10 @@ def _blueoil_to_lmnet(blueoil_config):
 
     learning_rate_func = None
     learning_rate_kwargs = None
-    if learning_rate_schedule == "constant":
+    if optimizer == "Momentum":
```
Comment (ruimashita) [inline blueoil/generate_lmnet_config.py]: It looks didn't care the case where "constant" and "Adam".

Comment (ruimashita) [reply blueoil/generate_lmnet_config.py]: When choose "constant" and "Adam", below are used.
```
             'LEARNING_RATE_FUNC': <function piecewise_constant at 0x7f0a11b89488>,
             'LEARNING_RATE_KWARGS': None,
             'OPTIMIZER_CLASS': <class 'tensorflow.python.training.adam.AdamOptimizer'>,
             'OPTIMIZER_KWARGS': {'learning_rate': 0.001},
```

`LEARNING_RATE_FUNC` should be set `None`

**Comment by ruimashita** [comment]
@iizukak 
Thanks, documentation is nice 👍 
I commented about handling optimizer and learning rate schedule, plz check it.

File: blueoil/generate_lmnet_config.py (hunk 4 above)
Comment (lm-konda) [reply blueoil/generate_lmnet_config.py]: @iizukak and I fixed this issue. Please review again.

File: blueoil/generate_lmnet_config.py [hunk 5]
```diff
@@ -155,9 +163,11 @@ def _blueoil_to_lmnet(blueoil_config):
     learning_rate_func = None
     learning_rate_kwargs = None
     if learning_rate_schedule == "constant":
-        optimizer_kwargs = {"momentum": 0.9, "learning_rate": initial_learning_rate}
+        if optimizer == "Momentum":
```
Comment (ruimashita) [inline blueoil/generate_lmnet_config.py]: `"momentum": 0.9` should be set when `optimizer` is `Mometum`, 
But `"momentum": 0.9` doesn't be set when `learning_rate_schedule` is not constant, I think.
Please check 👍

Comment (lm-konda) [reply blueoil/generate_lmnet_config.py]: You are right! I will fix it.

Comment (lm-konda) [reply blueoil/generate_lmnet_config.py]: Fixed and run blueoil_test.sh and run Adam + (constant / decay) and run Momentum + (constant / decay).

Comment (ruimashita) [reply blueoil/generate_lmnet_config.py]: When I use this config, `learning_rate_schedule` is `2-step-decay`
```
# supported task types are 'classification', 'object_detection' and 'semantic_segmentation'.
task_type: classification

network_name: LmnetV1Quantize

dataset:
  format: Caltech101
  train_path: ./lmnet/tests/fixtures/datasets/dummy_classification
  test_path: 

trainer:
  batch_size: 1
  epochs: 4
  # supported 'optimizer' is 'Momentum', 'Adam' currently.
  # Momentum
  #    https://www.tensorflow.org/api_docs/python/tf/train/MomentumOptimizer
  # Adam
  #    https://www.tensorflow.org/api_docs/python/tf/train/AdamOptimizer
  optimizer: Adam
  # supported 'learning_rate_schedule' is 'constant', '2-step-decay', '3-step-decay', '3-step-decay-with-warmup' ({epochs} is the number of training epochs you entered before).
  #   'constant' -> constant learning rate.
  #   '2-step-decay' -> learning rate decrease by 1/10 on {epochs}/2 and {epochs}-1.
  #   '3-step-decay' -> learning rate decrease by 1/10 on {epochs}/3 and {epochs}*2/3 and {epochs}-1.
  #   '3-step-decay-with-warmup' -> warmup learning rate 1/1000 in first epoch, then train the same way as '3-step-decay'.
  learning_rate_schedule: 2-step-decay
  initial_learning_rate: 0.001

network:
  quantize_first_convolution: yes

common:
  image_size:
    - 128  # height
    - 128  # width

  # set pretrain model name. currently, this feature is not supported, always ignored.
  pretrain_model: false

  # enable dataset prefetch, set false if weired problem happens
  dataset_prefetch: true

  data_augmentation:
    - Blur:
        - value: (0, 1)
```

But, Unfortunately, The learning rate is constant.

screen shot 2019-02-06 at 23 23 53


Need to fix 👍

Comment (lm-konda) [reply blueoil/generate_lmnet_config.py]: I confirmed with same settings.
The result is OK.

I think no fix is needed.

optimizer

Comment (lm-konda) [reply blueoil/generate_lmnet_config.py]: Sorry. The root of the problem is deep.
According to @nlpng 's help, setting both piecewise_constant and initilal_learning_value are not compatible.
I will try and confirm four(all) patterns.

Comment (lm-konda) [reply blueoil/generate_lmnet_config.py]: I confirmed four patterns and seems to work well. Please review again.
